### ISSUE TITLE
SQL injections

### ISSUE THREAD

**Issue by verekia** [knex/documentation#73]
Currently, the only place SQL injections are mentioned is the Raw section. It would be great to have a short section talking about this (or even one sentence somewhere). It is not clear whether Knex prevents or not SQL injections, and if there are any particular considerations regarding this topic when using Knex.

**Comment by lorensr** [comment]
Are there any places outside of `.raw` where SQL injection can occur? For example: 

```
.where('foo', 'like', `%${term}%`)
```

where `term` is provided by the client.

**Comment by elhigu** [comment]
There are some other places too, for example in schema API there are quite a few of SQL injection vulnerable places. 

`.where('foo', 'like', `%${term}%`)` this is safe, because it translates to 

```
sql: 'where "foo" like ?'
bindings: ['%userinput;tryingtohack%']
```

**Comment by dinis-cruz-gw** [comment]
Hi are those ‘other places where SQL injection are possible’ documented somewhere?

**Comment by elhigu** [comment]
Nope... but pretty much all the places in knex.schema.* methods and all the places where column / table identifiers are passed to the query are vulnerable. 

Thumb rule is that only the places where one passes values to SQL query like where(<identifier>, <value>) are protected. This is because value bindings are passed to query as `?` value replacements, which are then handled safely by the SQL driver. 

So to check if certain place is safe or not you can check `query.toSQL().sql` and if the passed value is rendered directly to that `.sql` string then it is not safe.

**Comment by MichaelBuhler** [comment]
This _needs_ to be documented.

**Comment by elhigu** [comment]
@MichaelBuhler PRs are welcome.

**Comment by acrosman** [comment]
Clarity on when Knex _intends_ for values to be sanitized and when it is left to the developer to provide protection would help people contribute documentation. I know I've run into places where values cause problems in DDL statements, but it is unclear when those are bugs and when that's expected behavior.  Certainly lots of developers _expect_ a library is good as Knex to sanitize inputs when possible or at least provide a consistent sanitizing function that can be called when needed, but the project intention here is unclear therefore making it hard to write good documentation.

**Comment by elhigu** [comment]
@acrosman Knex cannot provide any other protection to SQL injection, except for values that are passed to driver as bindings. So those are for example `.where('id', value)` where 'id' is outputted to final SQL as string and value is passed as parameter binding.

Those protected parts are easy to check by calling `.toSQL()` to your query and by looking what was generated and where bindings were used.   

> Certainly lots of developers expect a library is good as Knex to sanitize inputs when possible or at least provide a consistent sanitizing function that can be called when needed

Clearly you and lot's of developers doesn't understand how hard it is to write a such function for all the databases. 

It is so difficult task that it would deserve its own separate project in github, which then could be used for all database libraries in javascript world. I would also suppose that each DB driver would provide this kind of escaping function, but generally they doesn't seem to provide that.

If knex would try to provide something like that it would be filled with holes and it would provide  false sense of security for people using it and open even more SQL injection bugs to production code in the world. So knex is not even trying to say that it can  sanitize inputs, except for values that are passed as bindings to not providing that false sense of security.

**Comment by remram44** [comment]
A major problem is that there is no way to select a constant value. This is where I find myself using `select(knex.raw('...'))`, but then I have to escape the value before feeding it to `knex.raw()` to avoid injections.

This hole could be plugged if I could select from a binding instead.
